### PR TITLE
ci: split the full CI into two parts (running t/plugin and the other)

### LINKED ISSUE #5861: request help: split the full CI into two parts (running t/plugin and the other)
### Issue description

Sometimes the full CI "linux_openresty" https://github.com/apache/apisix/blob/4423509680a0c2103ce23b30a0da71735ef73e8f/.github/workflows/build.yml#L27
takes 50 minutes to complete.

Let's split it into multiple parts so the max single job time can be reduced.

We can pass an environment variable as the test file range in https://github.com/apache/apisix/blob/4423509680a0c2103ce23b30a0da71735ef73e8f/ci/linux_openresty_common_runner.sh#L86

### Environment

- apisix version (cmd: `apisix version`):
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### PR DESCRIPTION
**Pull request by tzssangglass**
Closes #5861

Signed-off-by: tzssangglass <tzssangglass@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -39,6 +39,14 @@ jobs:
         with:
           submodules: recursive
 
+      - name: Paths filter
+        uses: ./.github/actions/paths-filter
```
Comment (spacewander) [inline .github/workflows/build.yml:43]: Is it possible to remove this 3rd party action?

**Comment by spacewander** [comment]
> This PR doesn't solve the #5861. When non-plugin code is changed, we still need to run all the test cases.
> 
> I think we need to use matrix to run CI parallelly, and reduce the max run time.

Quoted from that issue:

> Let's split it into multiple parts so the max single job time can be reduced.

**Comment by tzssangglass** [comment]
> When non-plugin code is changed, we still need to run all the test cases.

yes, only the test files in t/plugin will be run if only the files in the plugins directory are modified.

> > Let's split it into multiple parts so the max single job time can be reduced.

Are you saying try to make the files in the apisix/ directory correspond to the files in the t/ directory? For example, if we modify the files in the `apisix/core` directory, then just run the test files under `t/plugin`.

I had considered doing this, and I was concerned that there wasn't enough coverage.

File: .github/workflows/build.yml (hunk 1 above)
Comment (tzssangglass) [reply .github/workflows/build.yml:43]: Then we need to implement a similar function ourselves, and I tried to find github action support for a similar function: trigger step by file modification path, but not found. I can't find the environment variable that indicates the file modification.

**Comment by kwanhur** [comment]
Multi-jobs run in parallel by default, see it [here](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobs). Splitting into two jobs(like build and build-plugin), then they'll cover all the scope together and reduce total cost time.

- build-plugin job proves `t/plugin` cases with preserving env variable value `true` or `on` etc

Hope this helps you.

**Comment by tzssangglass** [comment]
With the current split, in ci centos7 for example, `t/plugin` will take about 35 minutes and the other two will take about 20 minutes, I'm going to split `t/plugin` into two directories `t/plugin_1` and `t/plugin_2 `and run them in two jobs, what do you think? cc @spacewander @bzp2010 @shuaijinchao

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -104,4 +105,186 @@ jobs:
...
-        run: sudo ./ci/${{ matrix.os_name }}_runner.sh script
+        run: sudo -E ./ci/${{ matrix.os_name }}_runner.sh script
+
+
+  build-2:
+    strategy:
+      fail-fast: false
+      matrix:
+        platform:
+          - ubuntu-18.04
```
Comment (spacewander) [inline .github/workflows/build.yml:116]: Why not use matrix to solve the problem? We can use "test-dir" like how we use the `platform`

Comment (tzssangglass) [reply .github/workflows/build.yml]: fixed

**Review by spacewander — APPROVED** [review]

**Review by bzp2010 — APPROVED** [review]

**Pull request merged**